src/operator/control_flow.cc

```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+        }
+        if (i < (size_t) params.num_args) {
+          // a var
+          igrads[i] = (step == 0)
+                    ? outputs[i]
+                    : NDArray(outputs[i].shape(), outputs[i].ctx(), true, outputs[i].dtype());
+          iter_req[i] = (step == 0)
+                      ? req[i]
+                      : kWriteTo;
+          ++i;
+        }
+        else {
+          break;
+        }
+      }
+    }
+    state.Backward(step, ograds, iter_req, igrads);
+    for (size_t i = params.num_out_data; i < (size_t) params.num_args; ++i) {
+      size_t loc = params.func_var_locs[i - params.num_out_data];
+      ograds[i] = igrads[loc];
+    }
+  }
+  state.Cleanup();
+}
+
+static bool WhileLoopShape(const nnvm::NodeAttrs& attrs,
+                           std::vector<TShape> *in_shape,
+                           std::vector<TShape> *out_shape) {
+  using nnvm::ShapeVector;
+  const WhileLoopParam& params = nnvm::get<WhileLoopParam>(attrs.parsed);
+  // sanity checks
+  CHECK_EQ(in_shape->size() + 2U, (size_t) params.num_args);
+  CHECK_EQ(out_shape->size(), (size_t) params.num_outputs);
+  CHECK_EQ(attrs.subgraphs.size(), 2U);
+  CHECK_EQ(attrs.subgraphs[0]->outputs.size(), 1U);
+  // infer shape for cond and func
+  auto infer_subg = [&params, in_shape, out_shape](std::shared_ptr<Symbol> subg,
+                                                   ShapeVector *_subg_out,
+                                                   const nnvm::Tuple<dim_t> &input_locs,
+                                                   int num_out_data,
+                                                   bool fill_out_shape) {
+    // create subg_in
+    ShapeVector subg_in;
+    ShapeVector &subg_out = *_subg_out;
+    WhileLoopState::extract_by_loc(*in_shape, input_locs, &subg_in);
+    // create an indexed graph
+    nnvm::Graph g;
+    g.outputs = subg->outputs;
+    const auto& idx = g.indexed_graph();
+    // get input nodes
+    const auto &input_nids = idx.input_nodes();
+    // sanity checks
+    CHECK_EQ(input_nids.size(), subg_in.size());
+    CHECK_EQ(g.outputs.size(), subg_out.size());
+    CHECK_EQ(idx.input_nodes().size(), subg_in.size());
+    CHECK_EQ(idx.outputs().size(), subg_out.size());
+    // create empty shapes for inference
+    ShapeVector shapes(idx.num_node_entries());
+    // copy subg_in into shapes
+    for (size_t i = 0; i < subg_in.size(); ++i) {
+      auto eid = idx.entry_id(input_nids[i], 0);
+      shapes[eid] = subg_in[i];
+    }
+    // copy subg_out into shapes
+    // note that ndim of out_data is not increased
+    // because subg is only one step
+    for (size_t i = 0; i < subg_out.size(); ++i) {
+      auto eid = idx.entry_id(g.outputs[i]);
+      shapes[eid] = subg_out[i];
+    }
+    // copy done, call InferShape
+    g.attrs["shape"] = std::make_shared<dmlc::any>(std::move(shapes));
+    g = exec::InferShape(std::move(g));
+    if (g.GetAttr<size_t>("shape_num_unknown_nodes") != 0) {
```

Comment:
I think you need to copy the inferred shapes back first before returning. Shape inference takes multiple rounds. each round propagates shape info. having unknown shapes doesn't mean the shape inference fails.

Reply:
Please allow me to summarize it: after shape inference in a subgraph, we should do several things.
1) check consistency and possibly assign: `loop_vars` and `new_loop_vars`
2) check consistency and possibly assign: `subg_in` (shape of inputs of the subgraph) and `in_shape` (shape of inputs of the while_loop)

This is not limited to shape inference, I should also do it in type inference etc

Reply:
TODO List:
- [ ] shape inference
- [ ] type inference
- [ ] storage type inference

Reply:
solved